// @rspack/plugin-react-refresh/dist/index.js
import { createRequire } from "node:module";
import node_path from "node:path";
const d = (object, property, defaultValue)=>{
    if (void 0 === object[property] && void 0 !== defaultValue) object[property] = defaultValue;
    return object[property];
};
function normalizeOptions(options) {
    d(options, 'test', /\.(?:js|jsx|mjs|cjs|ts|tsx|mts|cts)$/);
    d(options, 'exclude', /node_modules/i);
    d(options, 'library');
    d(options, 'forceEnable', false);
    d(options, 'injectLoader', true);
    d(options, 'injectEntry', true);
    d(options, 'reloadOnRuntimeErrors', false);
    d(options, 'reactRefreshLoader', 'builtin:react-refresh-loader');
    return options;
}
const reactRefreshPath = node_path.join(import.meta.dirname, '../client/reactRefresh.js');
const reactRefreshEntryPath = node_path.join(import.meta.dirname, '../client/reactRefreshEntry.js');
const refreshUtilsPath = node_path.join(import.meta.dirname, '../client/refreshUtils.js');
const paths_require = createRequire(import.meta.url);
let refreshRuntimeDirPath;
function getRefreshRuntimeDirPath() {
    if (!refreshRuntimeDirPath) refreshRuntimeDirPath = node_path.dirname(paths_require.resolve('react-refresh/runtime'));
    return refreshRuntimeDirPath;
}
const getRefreshRuntimePaths = ()=>[
        reactRefreshEntryPath,
        reactRefreshPath,
        refreshUtilsPath,
        getRefreshRuntimeDirPath()
    ];
function addEntry(entry, compiler) {
    new compiler.rspack.EntryPlugin(compiler.context, entry, {
        name: void 0
    }).apply(compiler);
}
const PLUGIN_NAME = 'ReactRefreshRspackPlugin';
class ReactRefreshRspackPlugin {
    options;
    constructor(options = {}){
        this.options = normalizeOptions(options);
    }
    apply(compiler) {
        if (('development' !== compiler.options.mode || process.env.NODE_ENV && 'production' === process.env.NODE_ENV) && !this.options.forceEnable) return;
        if (this.options.injectEntry) addEntry(reactRefreshEntryPath, compiler);
        new compiler.rspack.ProvidePlugin({
            $ReactRefreshRuntime$: reactRefreshPath
        }).apply(compiler);
        if (this.options.injectLoader) compiler.options.module.rules.unshift({
            test: this.options.test,
            include: this.options.include,
            exclude: {
                or: [
                    this.options.exclude,
                    [
                        ...getRefreshRuntimePaths()
                    ]
                ].filter(Boolean)
            },
            resourceQuery: this.options.resourceQuery,
            dependency: {
                not: [
                    'url'
                ]
            },
            use: this.options.reactRefreshLoader
        });
        const definedModules = {
            __react_refresh_library__: JSON.stringify(compiler.rspack.Template.toIdentifier(this.options.library || compiler.options.output.uniqueName || compiler.options.output.library)),
            __reload_on_runtime_errors__: this.options.reloadOnRuntimeErrors
        };
        const providedModules = {
            __react_refresh_utils__: refreshUtilsPath
        };
        new compiler.rspack.DefinePlugin(definedModules).apply(compiler);
        new compiler.rspack.ProvidePlugin(providedModules).apply(compiler);
        compiler.options.resolve.alias = {
            'react-refresh': getRefreshRuntimeDirPath(),
            ...compiler.options.resolve.alias
        };
        compiler.hooks.compilation.tap(PLUGIN_NAME, (compilation)=>{
            compilation.hooks.additionalTreeRuntimeRequirements.tap(PLUGIN_NAME, (_, runtimeRequirements)=>{
                runtimeRequirements.add(compiler.rspack.RuntimeGlobals.moduleCache);
            });
        });
    }
}
export { ReactRefreshRspackPlugin };
