// react-render-to-markdown/dist/index.js
import react, { createContext } from "react";
import react_reconciler from "react-reconciler";
import { DefaultEventPriority, NoEventPriority } from "react-reconciler/constants.js";
class TextNode {
    text;
    parent;
    constructor(text){
        this.text = text;
    }
    setText(text) {
        this.text = text;
    }
}
class MarkdownNode {
    type;
    props;
    children;
    parent;
    constructor(type, props = {}){
        this.type = type;
        this.props = props;
        this.children = [];
    }
    appendChild(child) {
        this.children.push(child);
        if ('parent' in child) child.parent = this;
    }
    removeChild(child) {
        const index = this.children.indexOf(child);
        if (-1 !== index) this.children.splice(index, 1);
    }
    insertBefore(child, beforeChild) {
        const index = this.children.indexOf(beforeChild);
        if (-1 !== index) this.children.splice(index, 0, child);
        else this.appendChild(child);
        if ('parent' in child) child.parent = this;
    }
}
let currentUpdatePriority = NoEventPriority;
const hostConfig = {
    supportsMutation: true,
    supportsPersistence: false,
    supportsHydration: false,
    isPrimaryRenderer: false,
    getRootHostContext () {
        return {
            isInsideText: false
        };
    },
    getChildHostContext (parentHostContext, type) {
        const previousIsInsideText = parentHostContext.isInsideText;
        const isInsideText = 'text' === type || 'span' === type || previousIsInsideText;
        if (previousIsInsideText === isInsideText) return parentHostContext;
        return {
            isInsideText
        };
    },
    createInstance (type, props) {
        return new MarkdownNode(type, props);
    },
    createTextInstance (text) {
        return new TextNode(text);
    },
    shouldSetTextContent (_type, _props) {
        return false;
    },
    appendChild (parent, child) {
        parent.appendChild(child);
    },
    insertBefore (parent, child, beforeChild) {
        parent.insertBefore(child, beforeChild);
    },
    removeChild (parent, child) {
        parent.removeChild(child);
    },
    commitTextUpdate (textInstance, _oldText, newText) {
        textInstance.setText(newText);
    },
    commitUpdate (instance, _type, _oldProps, newProps, _internalHandle) {
        instance.props = newProps;
    },
    finalizeInitialChildren () {
        return false;
    },
    commitMount () {},
    clearContainer (container) {
        container.children = [];
    },
    setCurrentUpdatePriority (newPriority) {
        currentUpdatePriority = newPriority;
    },
    getCurrentUpdatePriority () {
        return currentUpdatePriority;
    },
    resolveUpdatePriority () {
        if (currentUpdatePriority !== NoEventPriority) return currentUpdatePriority;
        return DefaultEventPriority;
    },
    appendInitialChild (parent, child) {
        parent.appendChild(child);
    },
    preparePortalMount () {},
    resetTextContent () {},
    getPublicInstance (instance) {
        return instance;
    },
    prepareForCommit (_containerInfo) {
        return null;
    },
    resetAfterCommit (_containerInfo) {},
    scheduleTimeout: setTimeout,
    cancelTimeout: clearTimeout,
    noTimeout: -1,
    scheduleMicrotask: 'function' == typeof queueMicrotask ? queueMicrotask : (fn)=>Promise.resolve().then(fn),
    getInstanceFromNode () {
        return null;
    },
    beforeActiveInstanceBlur () {},
    afterActiveInstanceBlur () {},
    prepareScopeUpdate () {},
    getInstanceFromScope () {
        return null;
    },
    detachDeletedInstance () {},
    appendChildToContainer (container, child) {
        container.appendChild(child);
    },
    insertInContainerBefore (container, child, beforeChild) {
        container.insertBefore(child, beforeChild);
    },
    removeChildFromContainer (container, child) {
        container.removeChild(child);
    },
    hideInstance (_instance) {},
    hideTextInstance (textInstance) {
        textInstance.setText('');
    },
    unhideInstance (_instance) {},
    unhideTextInstance (textInstance, text) {
        textInstance.setText(text);
    },
    maySuspendCommit () {
        return false;
    },
    preloadInstance () {
        return true;
    },
    startSuspendingCommit () {},
    suspendInstance () {},
    waitForCommitToBeReady () {
        return null;
    },
    NotPendingTransition: null,
    HostTransitionContext: createContext(null),
    resetFormInstance () {},
    requestPostPaintCallback () {},
    shouldAttemptEagerTransition () {
        return false;
    },
    trackSchedulerEvent () {},
    resolveEventType () {
        return null;
    },
    resolveEventTimeStamp () {
        return -1.1;
    }
};
const reconciler = react_reconciler(hostConfig);
const ReactSharedInternals = react.__CLIENT_INTERNALS_DO_NOT_USE_OR_WARN_USERS_THEY_CANNOT_UPGRADE;
function noop() {}
let interceptorRefCount = 0;
let originalDescriptor;
let realH = null;
let cachedTarget = null;
let cachedProxy = null;
const transparentBlockTypes = new Set([
    'address',
    'article',
    'aside',
    'details',
    'div',
    'figcaption',
    'figure',
    'footer',
    'form',
    'header',
    'main',
    'nav',
    'section',
    'summary'
]);
const markdownBlockTypes = new Set([
    'blockquote',
    'h1',
    'h2',
    'h3',
    'h4',
    'h5',
    'h6',
    'hr',
    'li',
    'ol',
    'p',
    'pre',
    'table',
    'ul'
]);
function installEffectInterceptor() {
    if (!ReactSharedInternals) return noop;
    interceptorRefCount++;
    if (1 === interceptorRefCount) {
        originalDescriptor = Object.getOwnPropertyDescriptor(ReactSharedInternals, 'H');
        realH = ReactSharedInternals.H;
        cachedTarget = null;
        cachedProxy = null;
        Object.defineProperty(ReactSharedInternals, 'H', {
            get () {
                if (null == realH) return realH;
                if (cachedTarget !== realH) {
                    cachedTarget = realH;
                    cachedProxy = new Proxy(realH, {
                        get (target, prop, receiver) {
                            if ('useEffect' === prop || 'useLayoutEffect' === prop || 'useInsertionEffect' === prop) return noop;
                            return Reflect.get(target, prop, receiver);
                        }
                    });
                }
                return cachedProxy;
            },
            set (value) {
                realH = value;
            },
            configurable: true
        });
    }
    return ()=>{
        interceptorRefCount--;
        if (0 === interceptorRefCount) if (originalDescriptor) Object.defineProperty(ReactSharedInternals, 'H', originalDescriptor);
        else {
            ReactSharedInternals.H = void 0;
            ReactSharedInternals.H = realH;
        }
    };
}
function isFlowContainer(type) {
    return 'root' === type || transparentBlockTypes.has(type);
}
function isBlockBoundary(child) {
    return child instanceof MarkdownNode && (transparentBlockTypes.has(child.type) || markdownBlockTypes.has(child.type));
}
function startsWithNewline(value) {
    return 10 === value.charCodeAt(0);
}
function endsWithNewline(value) {
    return 10 === value.charCodeAt(value.length - 1);
}
function hasBlankLineBoundary(previousMarkdown, markdown) {
    return previousMarkdown.endsWith('\n\n') || markdown.startsWith('\n\n') || endsWithNewline(previousMarkdown) && startsWithNewline(markdown);
}
function getBlockSeparator(previousChild, child, previousMarkdown, markdown) {
    if (0 === previousMarkdown.length || 0 === markdown.length || !isBlockBoundary(previousChild) && !isBlockBoundary(child) || hasBlankLineBoundary(previousMarkdown, markdown)) return '';
    if (endsWithNewline(previousMarkdown) || startsWithNewline(markdown)) return '\n';
    return '\n\n';
}
function childToMarkdown(child) {
    if (child instanceof TextNode) return child.text;
    return toMarkdown(child);
}
function renderChildren(root) {
    const { children } = root;
    if (0 === children.length) return '';
    const shouldSeparateBlocks = isFlowContainer(root.type);
    const parts = [];
    let previousChild;
    let previousMarkdown = '';
    for (const child of children){
        const markdown = childToMarkdown(child);
        const separator = shouldSeparateBlocks && previousChild ? getBlockSeparator(previousChild, child, previousMarkdown, markdown) : '';
        if (separator) parts.push(separator);
        parts.push(markdown);
        if (markdown.length > 0) {
            previousChild = child;
            previousMarkdown = markdown;
        }
    }
    return parts.join('');
}
function toMarkdown(root) {
    const { type, props, children } = root;
    const getChildrenMarkdown = ()=>renderChildren(root);
    switch(type){
        case 'root':
            return getChildrenMarkdown();
        case 'h1':
            return `# ${getChildrenMarkdown()}\n\n`;
        case 'h2':
            return `## ${getChildrenMarkdown()}\n\n`;
        case 'h3':
            return `### ${getChildrenMarkdown()}\n\n`;
        case 'h4':
            return `#### ${getChildrenMarkdown()}\n\n`;
        case 'h5':
            return `##### ${getChildrenMarkdown()}\n\n`;
        case 'h6':
            return `###### ${getChildrenMarkdown()}\n\n`;
        case 'p':
            return `${getChildrenMarkdown()}\n\n`;
        case 'strong':
        case 'b':
            return `**${getChildrenMarkdown()}**`;
        case 'em':
        case 'i':
            return `*${getChildrenMarkdown()}*`;
        case 'code':
            if (root.parent?.type === 'pre') return getChildrenMarkdown();
            return `\`${getChildrenMarkdown()}\``;
        case 'pre':
            {
                const _language = props['data-lang'] || props.language || props.lang || '';
                const language = 'string' == typeof _language ? _language : '';
                const title = props['data-title'] || '';
                const block = [
                    'markdown',
                    'mdx',
                    'md',
                    ''
                ].includes(language) ? '````' : '```';
                return `\n${block}${language}${title ? ` title=${title}` : ''}\n${getChildrenMarkdown()}\n${block}\n`;
            }
        case 'a':
            return `[${getChildrenMarkdown()}](${props.href || '#'})`;
        case 'img':
            return `![${props.alt || ''}](${props.src || ''})`;
        case 'ul':
            return `${getChildrenMarkdown()}\n`;
        case 'ol':
            return `${getChildrenMarkdown()}\n`;
        case 'li':
            {
                const isOrdered = root.parent && 'ol' === root.parent.type;
                const prefix = isOrdered ? '1. ' : '- ';
                return `${prefix}${getChildrenMarkdown()}\n`;
            }
        case 'blockquote':
            return `> ${getChildrenMarkdown().split('\n').join('\n> ')}\n\n`;
        case 'br':
            return '\n';
        case 'hr':
            return '---\n\n';
        case 'style':
            return '';
        case 'table':
            return `${getChildrenMarkdown()}\n`;
        case 'thead':
            return getChildrenMarkdown();
        case 'tbody':
            return getChildrenMarkdown();
        case 'tr':
            {
                const cells = children.filter((child)=>child instanceof MarkdownNode).map((cell)=>toMarkdown(cell).trim());
                if (root.parent && 'thead' === root.parent.type) {
                    const separator = `|${cells.map(()=>' --- ').join('|')}|\n`;
                    return `| ${cells.join(' | ')} |\n${separator}`;
                }
                return `| ${cells.join(' | ')} |\n`;
            }
        case 'th':
        case 'td':
            return getChildrenMarkdown();
        default:
            return getChildrenMarkdown();
    }
}
async function renderToMarkdownString(element) {
    const container = new MarkdownNode('root');
    const root = reconciler.createContainer(container, 1, null, false, false, '', (error, info)=>{
        if (process.env.DEBUG) {
            console.error('Reconciler Error:', error, info);
            const message = `Reconciler Error:${error.message}`;
            throw new Error(message);
        }
    }, ()=>{}, (...args)=>{
        if (process.env.DEBUG) console.log('Reconciler onRecoverable Error:', ...args);
    }, ()=>{});
    const removeInterceptor = installEffectInterceptor();
    try {
        let resolveCommit = null;
        const commitPromise = new Promise((resolve)=>{
            resolveCommit = resolve;
        });
        reconciler.updateContainer(element, root, null, ()=>{
            if (resolveCommit) resolveCommit(toMarkdown(container));
        });
        return await commitPromise;
    } finally{
        removeInterceptor();
    }
}
export { renderToMarkdownString };
