// @rspress/core/dist/node/route/extractPageData.js
import { unified, node_path, promises, loadFrontMatter, remove, remark_gfm, MDX_OR_MD_REGEXP, remark_parse } from "../../740.js";
import { applyReplaceRules } from "../utils/applyReplaceRules.js";
import { flattenMdxContent } from "../utils/flattenMdxContent.js";
import { importStatementRegex } from "../constants.js";
import { parseToc } from "../mdx/remarkPlugins/toc.js";
function applyReplaceRulesToNestedObject(obj, replaceRules) {
    for(const key in obj)if ('string' == typeof obj[key]) obj[key] = applyReplaceRules(obj[key], replaceRules);
    else if ('object' == typeof obj[key] && null !== obj[key]) obj[key] = applyReplaceRulesToNestedObject(obj[key], replaceRules);
    return obj;
}
const remarkRemoveCodeBlocks = ()=>(tree)=>{
        remove(tree, 'code');
    };
const remarkRemoveImages = ()=>(tree)=>{
        remove(tree, 'image');
    };
const createProcessor = (searchCodeBlocks)=>unified().use(remark_parse).use(remark_gfm).use(remarkRemoveImages).use(searchCodeBlocks ? [] : [
        remarkRemoveCodeBlocks
    ]);
const processorWithCode = createProcessor(true);
const processorWithoutCode = createProcessor(false);
function extractTextFromNode(node) {
    if ('value' in node && 'string' == typeof node.value) return node.value;
    if ('children' in node && Array.isArray(node.children)) return node.children.map(extractTextFromNode).join('');
    return '';
}
const SEARCH_SKIP_TYPES = new Set([
    'image',
    'imageReference',
    'definition',
    'footnoteDefinition',
    'footnoteReference',
    'html',
    'thematicBreak',
    'mdxJsxFlowElement',
    'mdxJsxTextElement',
    'mdxFlowExpression',
    'mdxTextExpression',
    'mdxjsEsm'
]);
const SEARCH_BLOCK_TYPES = new Set([
    'paragraph',
    'heading',
    'listItem',
    'tableRow'
]);
function extractSearchText(node, codeblocks) {
    const { type } = node;
    if (SEARCH_SKIP_TYPES.has(type)) return [];
    if ('code' === type && !codeblocks) return [];
    if ('break' === type) return [
        '\n'
    ];
    if ('text' === type || 'inlineCode' === type) return [
        node.value
    ];
    if ('code' === type) return [
        `\n${node.value}\n`
    ];
    const result = [];
    if ('children' in node) for (const child of node.children)result.push(...extractSearchText(child, codeblocks));
    if (SEARCH_BLOCK_TYPES.has(type)) result.push('\n');
    if ('tableCell' === type) result.push('\t');
    return result;
}
function buildSearchContent(tree, rawToc, codeblocks) {
    const parts = [];
    const headingCharIndexes = [];
    let tocIndex = 0;
    let contentLength = 0;
    let skippedFirstH1 = false;
    for (const node of tree.children){
        if ('heading' === node.type && 1 === node.depth && !skippedFirstH1) {
            skippedFirstH1 = true;
            continue;
        }
        const text = extractSearchText(node, codeblocks).join('').replaceAll('\t\n', '\n').trim();
        if (text) {
            if (parts.length > 0) contentLength += 2;
            if ('heading' === node.type && node.depth >= 2 && node.depth < 5) {
                if (tocIndex < rawToc.length) {
                    headingCharIndexes.push(contentLength);
                    tocIndex++;
                }
            }
            contentLength += text.length;
            parts.push(text);
        }
    }
    return {
        content: parts.join('\n\n'),
        toc: rawToc.map((item, index)=>({
                ...item,
                charIndex: index < headingCharIndexes.length ? headingCharIndexes[index] : -1
            }))
    };
}
const SKIP_NODE_TYPES = new Set([
    'code',
    'html',
    'mdxjsEsm',
    'mdxFlowExpression',
    'thematicBreak',
    'image',
    'table'
]);
const CONTAINER_DIRECTIVE_REGEX = /^\s*:::\s*(\w+)\s*(.*)?/;
const CONTAINER_DIRECTIVE_END_REGEX = /^\s*:::\s*$/;
function stripContainerDirectives(text) {
    const lines = text.split('\n');
    const result = [];
    let directiveStartIndex = -1;
    for(let i = 0; i < lines.length; i++){
        if (-1 === directiveStartIndex && CONTAINER_DIRECTIVE_REGEX.test(lines[i])) {
            directiveStartIndex = i;
            continue;
        }
        if (-1 !== directiveStartIndex) {
            if (CONTAINER_DIRECTIVE_END_REGEX.test(lines[i])) directiveStartIndex = -1;
            continue;
        }
        result.push(lines[i]);
    }
    if (-1 !== directiveStartIndex) result.push(...lines.slice(directiveStartIndex + 1));
    return result.join('\n');
}
function extractDescription(tree) {
    const textParts = [];
    let foundH1 = false;
    for (const node of tree.children){
        if ('heading' === node.type && 1 === node.depth) {
            foundH1 = true;
            continue;
        }
        if ('heading' === node.type && 2 === node.depth) break;
        if (!SKIP_NODE_TYPES.has(node.type)) {
            if (foundH1) {
                const text = extractTextFromNode(node).trim();
                if (text) textParts.push(text);
            }
        }
    }
    const joined = textParts.join('\n');
    return stripContainerDirectives(joined).split('\n').map((l)=>l.trim()).filter(Boolean).join(' ');
}
async function getPageIndexInfoByRoute(route, options) {
    const { alias, replaceRules, root, searchCodeBlocks, extractDescription: extractDescriptionConfig = true, searchEnabled = true } = options;
    const defaultIndexInfo = {
        title: '',
        content: '',
        _flattenContent: '',
        routePath: route.routePath,
        lang: route.lang,
        toc: [],
        frontmatter: {},
        version: route.version,
        _filepath: route.absolutePath,
        _relativePath: node_path.relative(root, route.absolutePath).split(node_path.sep).join('/')
    };
    if (!MDX_OR_MD_REGEXP.test(route.absolutePath)) return defaultIndexInfo;
    let content = await promises.readFile(route.absolutePath, 'utf8');
    const { frontmatter, content: contentWithoutFrontMatter } = loadFrontMatter(content, route.absolutePath, root);
    applyReplaceRulesToNestedObject(frontmatter, replaceRules);
    const { flattenContent } = await flattenMdxContent(applyReplaceRules(contentWithoutFrontMatter, replaceRules), route.absolutePath, alias);
    content = flattenContent.replace(importStatementRegex, '');
    content = content.replace(/\r\n/g, '\n');
    const processor = searchCodeBlocks ? processorWithCode : processorWithoutCode;
    const tree = processor.parse(content);
    const { title, toc: rawToc } = parseToc(tree);
    const extractedDescription = frontmatter.description || !extractDescriptionConfig ? '' : extractDescription(tree);
    if (!searchEnabled) return {
        ...defaultIndexInfo,
        title: frontmatter.title || title,
        toc: rawToc.map((item)=>({
                ...item,
                charIndex: -1
            })),
        content: '',
        description: frontmatter.description || extractedDescription || void 0,
        _flattenContent: flattenContent,
        frontmatter: {
            ...frontmatter,
            __content: void 0
        }
    };
    const processedTree = await processor.run(tree);
    const { content: processedContent, toc } = buildSearchContent(processedTree, rawToc, searchCodeBlocks);
    return {
        ...defaultIndexInfo,
        title: frontmatter.title || title,
        toc,
        content: processedContent,
        description: frontmatter.description || extractedDescription || void 0,
        _flattenContent: flattenContent,
        frontmatter: {
            ...frontmatter,
            __content: void 0
        }
    };
}
async function extractPageData(routeService, options) {
    const pageData = await Promise.all(routeService.getRoutes().map(async (routeMeta)=>{
        const pageIndexInfo = await getPageIndexInfoByRoute(routeMeta, options);
        const routePage = routeService.getRoutePageByRoutePath(routeMeta.routePath);
        if (routePage) routePage.setPageIndexInfo(pageIndexInfo);
        return pageIndexInfo;
    }));
    return pageData;
}
var extractPageData_getPageIndexInfoByRoute = void 0;
export { extractPageData, extractPageData_getPageIndexInfoByRoute as getPageIndexInfoByRoute };
