// @rspress/core/dist/theme/components/Tabs/index.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { useStorageValue } from "@rspress/core/theme";
import clsx from "clsx";
import { Children, forwardRef, isValidElement, useMemo, useState } from "react";
import "./index.css";
function getTabValuesFromChildren(children, values) {
    if (values && values.length > 0) return values.map((item, index)=>{
        const content = children.find((child)=>child.props?.value === item.value) ?? children[index];
        return {
            ...item,
            content
        };
    });
    return Children.map(children, (child, index)=>{
        if (/*#__PURE__*/ isValidElement(child)) return {
            label: child.props?.label || void 0,
            value: child.props?.value,
            disabled: child.props?.disabled,
            content: children[index]
        };
        return {
            label: index,
            content: children[index]
        };
    });
}
const groupIdPrefix = 'rspress.tabs.';
const Tabs = /*#__PURE__*/ forwardRef((props, ref)=>{
    const { values, defaultIndex, defaultValue, onChange, children: rawChildren, groupId, tabPosition = 'left', className, labelItemClassName, contentItemClassName, keepDOM = true } = props;
    const children = Children.toArray(rawChildren).filter((child)=>!('string' == typeof child && '' === child.trim()) && /*#__PURE__*/ isValidElement(child));
    const tabValues = useMemo(()=>getTabValuesFromChildren(children, values), [
        children,
        values
    ]);
    if (import.meta.env.SSG_MD) return /*#__PURE__*/ jsx(Fragment, {
        children: tabValues.map(({ label, content })=>/*#__PURE__*/ jsxs(Fragment, {
                children: [
                    `\n**${label}**\n\n`,
                    content,
                    `\n`
                ]
            }))
    });
    const defaultValueIndex = void 0 !== defaultValue ? tabValues.findIndex((item)=>item.value === defaultValue) : -1;
    const initialActiveIndex = -1 === defaultValueIndex ? defaultIndex ?? 0 : defaultValueIndex;
    const [activeIndex, setActiveIndex] = useState(initialActiveIndex);
    const [storageIndex, setStorageIndex] = useStorageValue(`${groupIdPrefix}${groupId}`, initialActiveIndex.toString());
    const currentIndex = groupId ? Number(storageIndex) : activeIndex;
    return /*#__PURE__*/ jsxs("div", {
        className: clsx('rp-tabs', className),
        ref: ref,
        children: [
            tabValues.length ? /*#__PURE__*/ jsx("div", {
                className: "rp-tabs__label rp-tabs__label--no-scrollbar",
                style: {
                    justifyContent: 'center' === tabPosition ? 'center' : 'flex-start'
                },
                children: tabValues.map(({ label, value, disabled }, index)=>{
                    const isActive = index === currentIndex;
                    return /*#__PURE__*/ jsx("div", {
                        className: clsx('rp-tabs__label__item', isActive ? 'rp-tabs__label__item--selected' : 'rp-tabs__label__item--not-selected', labelItemClassName),
                        "data-index": index,
                        onClick: ()=>{
                            if (disabled) return;
                            onChange?.(index);
                            if (groupId) setStorageIndex(index.toString());
                            else setActiveIndex(index);
                        },
                        children: label
                    }, value ?? index);
                })
            }) : null,
            /*#__PURE__*/ jsx("div", {
                className: "rp-tabs__content",
                children: tabValues.map(({ value, content }, index)=>{
                    const isActive = index === currentIndex;
                    if (!keepDOM && !isActive) return null;
                    return /*#__PURE__*/ jsx("div", {
                        className: clsx('rp-tabs__content__item', isActive ? 'rp-tabs__content__item--active' : 'rp-tabs__content__item--hidden', contentItemClassName),
                        "aria-hidden": !isActive,
                        "data-index": index,
                        children: content
                    }, value ?? index);
                })
            })
        ]
    });
});
Tabs.displayName = 'Tabs';
function Tab({ children }) {
    return /*#__PURE__*/ jsx(Fragment, {
        children: children
    });
}
export { Tab, Tabs };
