// @rspress/core/dist/theme/assets/scroll-to-top.js
import { jsx, jsxs } from "react/jsx-runtime";
import "react";
const SvgScrollToTop = (props)=>/*#__PURE__*/ jsxs("svg", {
        xmlns: "http://www.w3.org/2000/svg",
        width: "1em",
        height: "1em",
        fill: "none",
        viewBox: "0 0 16 16",
        ...props,
        children: [
            /*#__PURE__*/ jsx("path", {
                stroke: "currentColor",
                strokeOpacity: 0.8,
                strokeWidth: 1.333,
                d: "M8 14.667A6.667 6.667 0 1 0 8 1.333a6.667 6.667 0 0 0 0 13.334Z",
                clipRule: "evenodd"
            }),
            /*#__PURE__*/ jsx("path", {
                stroke: "currentColor",
                strokeLinejoin: "round",
                strokeOpacity: 0.8,
                strokeWidth: 1.333,
                d: "M8 14.667A6.667 6.667 0 1 0 8 1.333a6.667 6.667 0 0 0 0 13.334Z"
            }),
            /*#__PURE__*/ jsx("path", {
                stroke: "currentColor",
                strokeLinecap: "round",
                strokeLinejoin: "round",
                strokeOpacity: 0.8,
                strokeWidth: 1.333,
                d: "M11 9 8 6 5 9"
            })
        ]
    });
var scroll_to_top = SvgScrollToTop;
export default scroll_to_top;
